A11y Addons control panel

deodorhunter opened the pull request:
This PR overhauls current Addons Control panel UI using react-aria-components and solves #5142. 
All accordions removed, UI is now akin to Plone Classic.
Aria pattern used: [ARIA grid pattern](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) through [GridList component](https://react-spectrum.adobe.com/react-aria/GridList.html#gridlist) 

**Changelog:**
- Written in Typescript leveraging `@plone/types`
- Component UI overhauled and code splitted, really minor changes to existing logic (types mostly)
- Updated tests and added new tests and related mocks/fixtures/snapshots for Cypress and Jest/Testing Library. 
- Updated some Addons related types in `@plone/types` package
- Added stylesheet and imported it directly in the component, css splitting and lazy loading works as expected. Will move styles in less theme files if deemed better/necessary
- Translations

**UI Preview**

![image](https://github.com/user-attachments/assets/d24679da-3231-4175-8a97-c86e17ad46e0)
![image](https://github.com/user-attachments/assets/9bca1271-8616-4204-a690-82a39e7ad2b1)

deodorhunter committed:
wip: rewriting addons control panel to be a11y compliant

deodorhunter committed:
feat: reworked addons control panel, added/fixed tests

deodorhunter committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

deodorhunter committed:
fix: added Changelog entry

deodorhunter committed:
fix: changelog entry name changed to reference pr instead of issue

deodorhunter committed:
fix: changelog entry for plone/types package

pnicolli committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

ichim-david commented on packages/volto/src/components/manage/Controlpanels/Addons/index.css:
```diff
@@ -0,0 +1,83 @@
+  padding: 2rem 1.5rem;
+}
+
+#page-addons .btn-primary {
```
I am not a fan of this styling of button variants where we introduce css overrides.
Since we are starting over and making use of css vars we should have a button class
and variant classes that where possible just modifies the vars.
Have a look at this article where this designer shows how they made their button
component.
https://piccalil.li/blog/how-i-build-a-button-component/#variants

I feel like we still need to adopt a styling methodology and try to use it in the new
components if we are now at the totally breaking future phase.
Will have another look at the plone.contents efforts and plone.components and
come back with further observations for the future.

pnicolli replied:
@ichim-david it's out of the scope of this PR to go in the direction of the new theming system. Since we don't know how that will be yet, it sounds like a waste of time for me to make it up in this specific PR. Moreover, Victor and I specifically told everybody working on reimplementing these panels and other views to avoid adding plone/components and such because that's not coming in volto 18 while these fixes should be there if possible.

deodorhunter replied:
@ichim-david I understand your concerns, and this kind of "custom temporary styling" is not my preferred style either.
I would have gone with a cva/cva like solution (https://cva.style/docs) if we had a theming convention (at least on paper), but alas that's still a long way in the future from what I understand.
As @pnicolli said, we had a talk about this during the sprint, and decided to avoid using plone/components and unified theming like quanta styles.

After some thinking, a possible edge case issue with my initial implementation of selector naming came to mind: people using custom cms-ui styles, like injecting bootstrap or a custom stylesheet with the same selector naming. It would have caused inconsistencies and hassles, so I changed the button selector names to reflect the scope of action of that button (install/install upgrade and uninstall).

deodorhunter committed:
fix: change css selectors naming and update all translations

deodorhunter committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

deodorhunter committed:
chore: update failing test snapshot due to classnames changes

deodorhunter commented:
**Update**
This pr is now complete with all missing translations in all supported languages. Got some help from our well known GPT buddy, as I would have hated leaving blanks for other languages.
Haven't seen strange/unusual translations during a quick cross check with Translate and Deepl, but you never know...

pnicolli committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

pnicolli committed:
Fixed tests and missing keys

ichim-david commented:
@pnicolli @sneridagh I've looked at the bundle implication of loading react-aria-components.
It seems that it's not bundled in the default bundle and only used on the addons
![react-aria-addons-component](https://github.com/user-attachments/assets/9969a2e0-89d8-4717-b149-ce5f2058741b)
I wonder it react-aria doesn't have better imports to avoid loading the entire library if you only use a couple of components.
Given the fact that we don't load it in the default bundle I'm not against this although I find it a bit strange that we have a dependency on react-aria-components within plone.components and now we load it also in volto core directly.
Whatever you guys decide I'm fine given that we don't fatten the default bundle :).

pnicolli committed:
Added h1 for a11y and fixed translations

pnicolli commented:
@ichim-david yeah Victor and I discussed this briefly with the people doing stuff with react-aria-components during Salamina Sprint and we said that we wouldn't use plone/components here because it's still alpha and kind of wip officially, but we can use react-aria-components for now, since moving from this to plone/components should remain easy enough.
Regarding a more precise import, we will need to look into it while working on plone/components so I think it's safe to say it's not a big problemi for now.

pnicolli committed:
Updated snapshot to fix tests

ichim-david reviewed (changes requested):
I see the following a11y issues:
1. Verify the warning given by IBM Equal Access Accessibility Checker plugin
2. I don't like that the screenreader says just listbox and you can only navigate with the arrows. Here it's strange if you use up and down arrow it will jump to the next listbox heading. If you use left or right keys it will read the heading, the install or uninstall button and the description such as "Installs the collective.volto.stickyblocks add-on." When you think of a listbox you think of up and down only so I don't know how many would think to use left or right to get to the button. Luckily it shows up in the form controls rotary so it could be worse.
3. The headings could use a bit of aria-label for instance what is available, if you just navigate through the headings you might not know what available means. 

![addons-headings](https://github.com/user-attachments/assets/76f01fdf-fb74-4e9a-ab35-d3875dc0847e)
![addons](https://github.com/user-attachments/assets/91e40e3a-f711-4b30-ac4a-0a25eace2ddc)

stevepiercy commented:
> I run checks with Accessible Web and Axe and no warning were shown, I'll try this addon too. For future reference, is this what the community usually uses as browser addon for a11y testing? I might have missed this piece of information, sorry if that's the case.

We don't document the tools we use, but we sure ought to.

https://6.docs.plone.org/volto/contributing/accessibility-guidelines.html

I'll update the [open PR](https://github.com/plone/volto/pull/6361/files) with the a11y tools that contributors should use (currently only cypress-axe is included), but please help me out by providing a list of their names (I've started that below), what you use them for, and from where to download or install them. Thank you!

* cypress-axe. Automate accessibility testing with [axe-core](https://github.com/dequelabs/axe-core) in [Cypress](https://cypress.io/).
* IBM Equal Access Accessibility Checker plugin.
* Accessible Web.
* Voiceover.
* moar tools

deodorhunter committed:
fix: review fixes and requested changes

ichim-david commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonItem.tsx:
```diff
@@ -83,7 +112,8 @@ const AvailableItem: React.FC<AvailableAddonProps> = ({ addon, onInstall }) => {
       key={addon['@id']}
       className="addon-item"
-      aria-describedby={`addon-desc-${addon.id}`}
+      textValue={`${addon.title} ${addon.description}. ${intl.formatMessage({ id: 'Press Enter to install this addon' })}`}
```
@deodorhunter you add the period which shows up strange if you don't have an addon. See the test result here https://github.com/plone/volto/pull/6345/commits/416059d6f1cbe1423d83a68249cada263504c4f4#diff-023c7e608404538c6858332e1915002b4c248049367176e7d0a52f69097794a2R65

deodorhunter committed:
fix: remove aria-label period breaking unrelated slate page test in metadata.js somehow

deodorhunter committed:
chore: update jest snapshot

deodorhunter committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

deodorhunter committed:
chore: realign .pot files after main merge

pnicolli commented:
@ichim-david Not sure I understand the status today. Are the review comments still open or are they already settled? Are there any blockers among those that are still open? Just trying to figure out if this will make it into v18 or not and how to plan for the future.

ichim-david commented:
> @ichim-david Not sure I understand the status today. Are the review comments still open or are they already settled? Are there any blockers among those that are still open? Just trying to figure out if this will make it into v18 or not and how to plan for the future.

I  have a few more comments that I will add soon. This is an improvement over the previous add-ons control panel so I am not against having it for Volto 18. Overall it's just minor issues left that can be handled. @sneridagh should also give his input on this work.

ichim-david reviewed (changes requested):
@deodorhunter @pnicolli  there is no Upgradable, it's Upgradeable. Please rename everywhere in order for it to sound proper when using a screenreader. This also means every import and file name that has Upgradable.

Also please use add-ons instead of addons as it will be read again better by screen readers

ichim-david commented:
> @deodorhunter @pnicolli there is no Upgradable, it's Upgradeable. Please rename everywhere in order for it to sound proper when using a screenreader. This also means every import and file name that has Upgradable.
> 
> Also please use add-ons instead of addons as it will be read again better by screen readers

@deodorhunter ok I was incorrect to mention that there is no upgradable, there is https://en.wiktionary.org/wiki/upgradable#English. However I use American english and not british english and it sounded better for me with upgradeable. I don't know if this is reason enough to change it though, @stevepiercy do you have anything to add on this?

ichim-david commented:
@deodorhunter @pnicolli I would also do the following fixes:
- align the buttons to the top
- add a flex-wrap on the header
- add a class to the div that has the install, update or delete button
![addons-320px-flex-wrap](https://github.com/user-attachments/assets/340e6287-1465-4b2c-8bb0-54d2bcef7b89)
![addons-320px](https://github.com/user-attachments/assets/b25a5f7b-3dc5-4402-a703-ef8cdc48a8c0)
![addons-360px](https://github.com/user-attachments/assets/2f126999-9b4d-4004-9c59-46ea0c0fe18a)

ichim-david commented:
@deodorhunter I don't like the behavior of double-clicking on the section which will trigger an install or uninstall. it's not at all obvious and is in fact totally unexpected. I might want to copy some of the text from that section and install I get an install or uninstall

https://github.com/user-attachments/assets/aa48e8ed-a363-4c68-a202-b713d8b721c5

I see that you mention something about double-clicking performing an action https://github.com/plone/volto/pull/6345#issuecomment-2446936473.
I find this change worse than what we previously had with the left and right arrows. If you can't quickly do something about it I would still disable the double click to perform action.

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -434,10 +434,15 @@ msgid "Automatically"
 
+#. Default: "Available addons"
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "Available addons"
```
```suggestion
msgid "Available add-ons"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -1871,22 +1876,27 @@ msgid "Inspect relations"
 
+#. Default: "Installed addons"
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "Installed addons"
```
```suggestion
msgid "Installed add-ons"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2382,6 +2392,16 @@ msgstr ""
 
+#. Default: "No addons available for installation."
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "No Addon available"
```
```suggestion
msgid "No add-on available"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2382,6 +2392,16 @@ msgstr ""
+
+#. Default: "No installed addons."
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "No Installed available"
```
```suggestion
msgid "No installed add-ons available"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2382,6 +2392,16 @@ msgstr ""
+msgid "No Addon available"
+msgstr ""
+
+#. Default: "No installed addons."
```
```suggestion
#. Default: "No installed add-ons."
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -255,32 +255,32 @@ msgid "Additional date"
 
-#. Default: "Addon uninstalled succesfuly"
-#: components/manage/Controlpanels/AddonsControlpanel
+#. Default: "Addon uninstalled successfully"
```
```suggestion
#. Default: "Add-on uninstalled successfully"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -255,32 +255,32 @@ msgid "Additional date"
 
-#. Default: "Addon installed succesfuly"
-#: components/manage/Controlpanels/AddonsControlpanel
+#. Default: "Addon installed successfully"
```
```suggestion
#. Default: "Add-on installed successfully"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2746,6 +2771,21 @@ msgstr ""
 msgid "Preferences"
 msgstr ""
 
+#. Default: "Press Enter to install this addon"
```
```suggestion
#. Default: "Press Enter to install this add-on"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2746,6 +2771,21 @@ msgstr ""
 
+#. Default: "Press Enter to install this addon"
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "Press Enter to install this addon"
```
```suggestion
msgid "Press Enter to install this add-on"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2746,6 +2771,21 @@ msgstr ""
+msgid "Press Enter to install this addon"
+msgstr ""
+
+#. Default: "Press Enter to uninstall this addon"
```
```suggestion
#. Default: "Press Enter to uninstall this add-on"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2746,6 +2771,21 @@ msgstr ""
+msgid "Press Enter to uninstall this addon"
+msgstr ""
+
+#. Default: "Press Enter to update this addon"
```
```suggestion
#. Default: "Press Enter to update this add-on"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2746,6 +2771,21 @@ msgstr ""
+
+#. Default: "Press Enter to update this addon"
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "Press Enter to update this addon"
```
```suggestion
msgid "Press Enter to update this add-on"
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -3812,6 +3852,11 @@ msgstr ""
 
+#. Default: "This addon was updated. Current profile installed version is {installedVersion}. New available profile version is {newVersion}"
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "This addon was updated. Current profile installed version is {installedVersion}. New available profile version is {newVersion}"
```
```suggestion
msgid "This add-on was updated. Current profile installed version is {installedVersion}. New available profile version is {newVersion}."
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -4069,21 +4114,21 @@ msgstr ""
 
+#. Default: "Upgradable addons"
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "Upgradable addons"
```
```suggestion
msgid "Upgradable add-ons"
```

stevepiercy commented on packages/volto/news/6345.bugfix:
```diff
@@ -0,0 +1 @@
+Overhaul Addons control panel with accessible UI (react-aria-components) akin to Plone Classic @deodorhunter
```
```suggestion
Overhaul Add-ons control panel with accessible user interface (react-aria-components) akin to Plone Classic UI. @deodorhunter
```

stevepiercy commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonsControlpanel.tsx:
```diff
@@ -0,0 +1,356 @@
+              />
+              &nbsp;
+              <a
+                href="https://6.docs.plone.org/install/"
```
Hold on this. We are in the process of updating the documentation, and the new page is not yet released. See https://github.com/plone/volto/pull/6397/files#diff-5ac1c30eabae6878f93007dc64b3129cc05eaa5f878cd274fc27779edd410276

```suggestion
                href="https://6.docs.plone.org/volto/development/add-ons/install-an-add-on.html"
```

stevepiercy reviewed (changes requested):
In my review the changes in `volto.pot` are authoritative and should be propagated to the other files.

stevepiercy commented:
@ichim-david thanks for the ping.

"add-on" is hyphenated in narrative documentation and any public UI element. "addon" is preferred in code. Where the spelling of "successfully" was corrected, I also corrected the hyphenation of "add-on", but only in `volto.pot`, as I think the others are auto-generated, except for `*.tsx` files. In my review the changes in `volto.pot` are authoritative and should be propagated to the other files. Please propagate those suggestions.

However, I don't know if changing all the non-hyphenated instances of "addon" in the `msgid`s would be a serious breaking change. Can someone inform me?

In any case, it bugs me that we don't use the correct hyphenation or spelling when displaying the text to end users. Perhaps we deal with this misspelling (and probably others that I never reviewed) in a future PR and 18.1 release?

"Upgradeable" is a variant spelling of "upgradable" in American English. Both spellings are acceptable and interchangeable.

ichim-david commented:
> @ichim-david Not sure I understand the status today. Are the review comments still open or are they already settled? Are there any blockers among those that are still open? Just trying to figure out if this will make it into v18 or not and how to plan for the future.

@pnicolli given Victor announcement that a48 is the final this for sure won't make it in 18 final but having it in 18.1 means that you get more feedback besides my own, I know that @JeffersonBledsoe also wanted to add something but cannot do it until tomorrow.

stevepiercy commented:
> If you change it you lose the translations and you need to add them back manually. I added a bug report here #6159 but that was as far as I got.

In that case, let's not change existing `addon`, but we should change new `addon` to `add-on`, just as I did in my review. We can address existing misspellings in the future, if we can figure out how to do so without being destructive.

pnicolli commented:
> @pnicolli given Victor announcement that a48 is the final this for sure won't make it in 18 final but having it in 18.1 means that you get more feedback besides my own, I know that @JeffersonBledsoe also wanted to add something but cannot do it until tomorrow.
@ichim-david yeah after reading replies here I briefly texted Victor to pull the trigger without waiting for this. I am just afraid this will become stale if we go too deep in optimization. I would like to set a goal to have this finished by the end of the month at the latest.

davisagli commented:
>> If you change it you lose the translations and you need to add them back manually. I added a bug report here https://github.com/plone/volto/issues/6159 but that was as far as I got.

> In that case, let's not change existing addon, but we should change new addon to add-on, just as I did in my review. We can address existing misspellings in the future, if we can figure out how to do so without being destructive.

We can change the defaultMessage of existing translations but leave the msgid the same.

stevepiercy commented:
> Sure thing, a couple that come to mind and usually use:

Added in 21505f5c18142329aa9536ac5f0a4ebb94afb407.

deodorhunter committed:
fix: revert double click to install action, minor css fixes, replace all occurrences of addon with add-ons in translation and intl messages

deodorhunter committed:
fix: more intl messages and translations fixes for 'addon'

deodorhunter committed:
chore: realign to main and update test snapshots

deodorhunter committed:
fix: revert git merge and realign again, recover missing files after merge

deodorhunter committed:
fix: restore old translations due to known translations bug, restore missing files due to git merging with wrong strategy

deodorhunter committed:
chore: align to main and fix left in tree conflict from merge, fix new linting issues after main alignment

deodorhunter committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

ichim-david commented:
@deodorhunter your pull request is out of sync, after bringing it up to date let me know if everything is ready for the second time review.

deodorhunter committed:
chore: bridge branch for merging

deodorhunter committed:
fix: wrong merge of docs/source/release-notes/index.md

deodorhunter committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

deodorhunter commented:
> @deodorhunter your pull request is out of sync, after bringing it up to date let me know if everything is ready for the second time review.
I had a couple busy months, sorry for leaving this hanging.
It is now up to date with main and ready for review again, CI fails on changelog check but I'm struggling to understand why, news files of changed code in this pr are included

stevepiercy commented:
Here's a clue https://github.com/plone/volto/actions/runs/13967359876/job/39100719196?pr=6345#step:11:3

```
towncrier check --compare-with origin/main --dir packages/generator-volto
```

That means towncrier looked in `packages/generator-volto` for a news item, but found none, because of a change in this file https://github.com/plone/volto/pull/6345/files#diff-918b4e82877daa8aee62ce41977634a3a267f26f503562e1355c416ff643713b.

deodorhunter committed:
fix: missing news file in packages/generator-volto

deodorhunter commented:
> Here's a clue https://github.com/plone/volto/actions/runs/13967359876/job/39100719196?pr=6345#step:11:3
> 
> ```
> towncrier check --compare-with origin/main --dir packages/generator-volto
> ```
> 
> That means towncrier looked in `packages/generator-volto` for a news item, but found none, because of a change in this file https://github.com/plone/volto/pull/6345/files#diff-918b4e82877daa8aee62ce41977634a3a267f26f503562e1355c416ff643713b.

Thank you so much for the hint, my brain was really skipping that one find and replace change in the diff, all tests should be green now

stevepiercy commented:
Something doesn't look right in the results after the merge. Why are files edited that should not be edited, including news files from PRs that were merged and released months ago? I'll hold my review for now.

- https://github.com/plone/volto/pull/6345/files#diff-8ea15a25d044c7930c6e3f0f521299f7b93d2d8196a08ac0e1c4898f73869d5f
- https://github.com/plone/volto/pull/6345/files#diff-b5aa808176801bfde2277ac886bbbdf0863de109950303cdb89e2d33ebe16a28

deodorhunter committed:
fix: remove unwanted formatting changes from .md linter after main merge

deodorhunter committed:
fix: merge conflicts from main

deodorhunter committed:
fix: ru translation after git merge fail

stevepiercy commented:
The failing linkcheck is for https://www.npmjs.com/search?q=keywords%3Avolto-addon%2Cvolto, but it should be only temporary and can be ignored.

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -85,8 +85,8 @@ msgid "Add (object list)"
-#: components/manage/Controlpanels/AddonsControlpanel
-msgid "Add Addons"
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "Add Add-ons"
```
Existing `msgid`s should not be changed, even though they are not correctly hyphenated, as it may break stuff. See https://github.com/plone/volto/issues/6159. Please revert any existing `msgid`s that you changed.

You may change `Default` or `msgstr`, or add or edit components.

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2423,14 +2443,19 @@ msgstr ""
 msgid "No Transactions Selected To Do Undo"
 msgstr ""
 
+#. Default: "No upgrades in this corner. You are up to date. High fives."
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "No Upgrades available"
+msgstr ""
+
 #. Default: "No Video selected"
 #: components/manage/Blocks/Video/VideoSidebar
 msgid "No Video selected"
 msgstr ""
 
-#. Default: "No addons found"
+#. Default: "No add-ons found"
 #: components/manage/Controlpanels/VersionOverview
-msgid "No addons found"
+msgid "No add-ons found"
 msgstr ""
```
This is a good example of where you it is OK to change the value for `Default` but should revert the change in value for `msgid`.

stevepiercy commented on packages/volto/package.json:
```diff
@@ -365,6 +366,7 @@
     "razzle": "4.2.18",
     "razzle-dev-utils": "4.2.18",
     "razzle-plugin-scss": "4.2.18",
+    "react-aria-components":"^1.5.0",
```
Shouldn't the dependencies versions match?

stevepiercy reviewed (changes requested):
There's still some news files in here that have already been merged and released that shouldn't show up. Something isn't right with this branch's git history.

deodorhunter committed:
fix: remove merged news files and older react-aria-components in deps

deodorhunter committed:
fix: remove merged news files

deodorhunter committed:
fix: try find and replace of changed msgids due to volto bug

deodorhunter committed:
fix: missing italian translation

deodorhunter committed:
fix: trying to restore 7mo history in translations for hyphenated addon message id

deodorhunter commented:
> There's still some news files in here that have already been merged and released that shouldn't show up. Something isn't right with this branch's git history.

This pr is pre volto 18 (Salamina Sprint), was supposed to me merged for volto 18 but sadly fell through. It was a nightmare to get it up to date, git failed me pretty hard on a lot of left in tree and our-theirs cases. Last clean up should have smoothed things out (thanks for helping!), I'm keeping commits atomic to be able to roll back easily if need arises, 7 months have passed and I remember almost nothing 😅

deodorhunter committed:
fix: italian translations

deodorhunter committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

pnicolli committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

pnicolli commented:
@plone/volto-accessibility did you find some time to look into this? We should try to avoid having PRs that are ready and become stale.

JeffersonBledsoe commented:
@pnicolli Thanks for the re-ping for the accessibility team. I'll take a review next week and follow up with the others!

Wagner3UB commented:
@deodorhunter an you give me an update on this PR, please? Do you think it can be finished, or at least merged with some improvements? Thank you!

deodorhunter commented:
> @deodorhunter an you give me an update on this PR, please? Do you think it can be finished, or at least merged with some improvements? Thank you!

Hi @Wagner3UB! 
So, PR was complete and potentially ready to be merged with fixes from previous reviews implemented a long time ago. Sadly, we are looking at a 1+ year drift on this, my brain no longer has a working memory of the domain/scope. 
Aside from some tedious merge conflicts solving as shown by GitHub (last time I worked on Volto was v18), could you be more specific on what kind of improvements would we need on this one? Did you test it accessibility wise and found some issue?
IIRC this was built on a much earlier version of React Aria Components, we should also probably check if improvements or breaking changes were made to the lib and components themselves during the hiatus.
If I have a clear, defined scope of what needs changes/fixes/improvements I can take a look at this in the coming weeks :)

pnicolli commented:
@Wagner3UB I remember that this was indeed finished and just needed a review from the a11y team which was postponed but it makes sense to review and merge this now that the team regained momentum and energy. I would recommend solving the conflicts and reviewing as-is, I don't think it needs functional changes (unless it has a11y bugs ofc)

Wagner3UB commented:
Great! Thanks for the clarification @deodorhunter and @pnicolli.

I didn’t go beyond reading the whole discussion yet, but based on your feedback, we can definitely move forward. I’m also happy to hear that it is already finished.

Marty, no worries, I can handle the conflict resolution myself. I’m a bit more familiar with the current context and should be able to do it more quickly. I’ll also run some tests and then ask the a11y team members to take a general look at it.

Thanks!

Wagner3UB committed:
Merge branch 'main' into 5142-a11y---site-setup---addon-ons---cms-ui

# Conflicts:
#	packages/volto/locales/ca/LC_MESSAGES/volto.po
#	packages/volto/locales/de/LC_MESSAGES/volto.po
#	packages/volto/locales/en/LC_MESSAGES/volto.po
#	packages/volto/locales/es/LC_MESSAGES/volto.po
#	packages/volto/locales/eu/LC_MESSAGES/volto.po
#	packages/volto/locales/fi/LC_MESSAGES/volto.po
#	packages/volto/locales/fr/LC_MESSAGES/volto.po
#	packages/volto/locales/hi/LC_MESSAGES/volto.po
#	packages/volto/locales/it/LC_MESSAGES/volto.po
#	packages/volto/locales/ja/LC_MESSAGES/volto.po
#	packages/volto/locales/nl/LC_MESSAGES/volto.po
#	packages/volto/locales/pt/LC_MESSAGES/volto.po
#	packages/volto/locales/pt_BR/LC_MESSAGES/volto.po
#	packages/volto/locales/ro/LC_MESSAGES/volto.po
#	packages/volto/locales/ru/LC_MESSAGES/volto.po
#	packages/volto/locales/volto.pot
#	packages/volto/locales/zh_CN/LC_MESSAGES/volto.po
#	packages/volto/package.json
#	packages/volto/src/components/manage/Controlpanels/Addons/AddonsControlpanel.test.jsx
#	packages/volto/src/components/manage/Controlpanels/AddonsControlpanel.jsx
#	packages/volto/src/components/manage/Controlpanels/__snapshots__/AddonsControlpanel.test.jsx.snap
#	pnpm-lock.yaml

Wagner3UB committed:
chore: translate new addons msgids and restore lost upgradeVersions entry

Translate the 7 new msgids introduced by the Addons control panel
refactor across all locales, and restore the "upgradeVersions" msgid
that was silently dropped from volto.pot during the merge from main,
which had cascaded into ca, eu and pt losing their existing translation.

Wagner3UB committed:
fix: declare defaultMessage for upgradeVersions to keep it in i18n extraction

The intl.formatMessage call referenced the 'upgradeVersions' id without
a defaultMessage, so the official pnpm i18n extraction silently dropped
it from volto.pot on every run, wiping the translation from all locales.

Wagner3UB committed:
fix: restore keyboard access to addon install/uninstall/upgrade actions

Wagner3UB committed:
fix: add screen reader instructions for install/uninstall/update actions

Wagner3UB committed:
fix: add navigation hint to first item in addons list for screen readers

Wagner3UB committed:
fix: correct AddonsControlpanel Props typing for build:types

Wagner3UB commented:
@deodorhunter @pnicolli 

I went through the merge conflicts, standardizing some files in the process, but kept the original structure and intent intact. I also updated a few dependency versions that were out of sync, since `main` had moved on quite a bit since this branch was last rebased.

While testing the merged result, I found and fixed a few accessibility issues:

- **Install/uninstall/update was unreachable by keyboard (pre-existing).** Every row had `isDisabled` set unconditionally, left over from an earlier design where the row itself triggered the action, before it was reworked to use a dedicated button. With `isDisabled` set, Tab and Enter/Space did nothing, only a mouse click worked. Removed `isDisabled` and wired `onAction` on each row, so Enter/Space now triggers the same action as the button.
- **The "update" label was reading raw placeholders (pre-existing).** The `upgradeVersions` message had no `defaultMessage`, so `pnpm i18n` silently dropped it from every locale file. The button's aria-label read literally "Update from version {origin} to {destination}" instead of real version numbers. Fixed by declaring the message properly and passing the actual values.
- **The add-ons list rendered as a tiny scrollable box instead of filling the section (caused by the merge).** `@plone/components`, a dependency that only exists on `main`, ships a generic base style for `.react-aria-GridList` (`width: 250px`, `max-height: 300px`), meant for compact widgets like dropdowns. Since `react-aria-components` always applies that class name unless overridden, it leaked into our list once the branches merged. Fixed by giving each `GridList` its own `className`.
- `main` also brought in a conflicting `react-router` version, which broke `build:types` in CI with an ambiguous `RouteComponentProps` type. Fixed by typing the one prop actually used (`location.pathname`) directly instead of relying on that import.

Examples:
First item list:
<img width="1078" height="607" alt="first-item-list" src="https://github.com/user-attachments/assets/8b038e29-57a6-4130-beb5-8e7b237ac704" />

Second item list:
<img width="1412" height="507" alt="second-item-list" src="https://github.com/user-attachments/assets/a318b828-2637-4f00-9aeb-4ee557b538ae" />

Wagner3UB commented:
One thing I want to flag for discussion. The component is built on a `GridList` (which comes with `role="grid"`, i.e. rows and columns), but in practice the layout is always a single column with several rows, and that's not going to change. For someone using a screen reader, this makes the widget announce itself as a full 2D table ("table, 1 column, N rows", plus the "press Control-Option-Shift-Down to enter this table" prompt), way more complexity than the content actually has. It's not a table, it's a list of buttons, each tied to one action: install or uninstall.

What I'd expect instead: a labeled container per section ("Updates available: 0", "Available add-ons: 6", "Installed add-ons: 3"), and inside it a plain list of buttons, each with its own clear label. If some structure is still needed, Enter could get you into the container and Tab could move between items, without the table semantics on top. This would likely mean switching from `GridList` to `ListBox` (both from `react-aria-components`), which I didn't do here since it touches more of the component's structure and felt like a bigger call to make unilaterally. Happy to take it on if the team agrees it's the right direction.

cc:/ @deodorhunter @pnicolli @wesleybl @Humanaice

Wagner3UB committed:
fix: remove duplicate react-aria-components entry and apply review wording suggestions

Wagner3UB committed:
fix: rename new addons msgids per review suggestions

Wagner3UB commented:
Quick follow-up on the open points above:

- **Double-click**: verified, it doesn't trigger an install/uninstall on its own. A single click does its job once, and I couldn't get an extra call out of it even forcing a double-click event directly.
- **Narrow viewports (320px and below)**: tested down to 301px, the layout holds up fine. Title wraps to a second line when needed and the button stays put, no overlap.
- **`GridList` vs `ListBox`**: leaving this as `GridList` for now, per the discussion above. Not blocking this PR.

stevepiercy commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonsControlpanel.tsx:
```diff
@@ -0,0 +1,368 @@
+  addAddons: {
+    id: 'Add Addons',
+    defaultMessage:
+      'To make new add-ons show up here, add them to your configuration, build, and restart the server process. For detailed instructions see',
```
see... what?

stevepiercy commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonsControlpanel.tsx:
```diff
@@ -0,0 +1,368 @@
+              </h2>
+              <FormattedMessage
+                id="Add Addons"
+                defaultMessage="To make new add-ons show up here, add them to your configuration, build, and restart the server process. For detailed instructions see"
```
see... what?

Wagner3UB replied:
The old structure was 'text i18n' + 'hardcoded HTML link', which is why the translations seemed incomplete. Now I've passed the link via a prop to i18n.

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2577,6 +2597,11 @@ msgstr ""
 msgid "No images found."
 msgstr ""
 
+#. Default: "No installed add-ons."
+#: components/manage/Controlpanels/Addons/AddonsControlpanel
+msgid "No installed add-ons available"
```
Between the two, I'm not sure which is more correct, so I'll change both so they match each other.

```suggestion
#. Default: "No installed add-ons available."
#: components/manage/Controlpanels/Addons/AddonsControlpanel
msgid "No installed add-ons available."
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -2931,6 +2956,21 @@ msgstr ""
 
+#. Default: "Press Enter to install this add-on."
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "Press Enter to install this add-on"
+msgstr ""
+
+#. Default: "Press Enter to uninstall this add-on."
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "Press Enter to uninstall this add-on"
+msgstr ""
+
+#. Default: "Press Enter to update this add-on."
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "Press Enter to update this add-on"
```
This is nitpicky for consistency, but probably a good thing to do.

```suggestion
msgid "Press Enter to install this add-on."
msgstr ""

#. Default: "Press Enter to uninstall this add-on."
#: components/manage/Controlpanels/Addons/AddonItem
msgid "Press Enter to uninstall this add-on."
msgstr ""

#. Default: "Press Enter to update this add-on."
#: components/manage/Controlpanels/Addons/AddonItem
msgid "Press Enter to update this add-on."
```

stevepiercy commented on packages/volto/locales/volto.pot:
```diff
@@ -4386,6 +4431,11 @@ msgstr ""
 
+#. Default: "Use the arrows to move between items."
+#: components/manage/Controlpanels/Addons/AddonItem
+msgid "Use the arrows to move between items"
```
Another nitpick.

```suggestion
msgid "Use the arrows to move between items."
```

stevepiercy reviewed (changes requested):
The `id` `Add Addons` doesn't have a matching `msgid`. Also its `defaultMessage` is incomplete.

wesleybl commented on packages/types/src/services/addons.d.ts:
```diff
@@ -1,17 +1,25 @@
+  hasProfile: boolean;
+  installedVersion: string;
+  newVersion: string;
+  requires: boolean;
```
It is `required`.

wesleybl commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonItem.tsx:
```diff
@@ -0,0 +1,188 @@
+      textValue={
+        (addon.upgrade_info.installedVersion && addon.upgrade_info.newVersion
+          ? `${addon.description} ${descriptionText}`
+          : addon.description + addon.upgrade_info.available) +
```
When `installedVersion` or `newVersion` are missing, the fallback is `addon.description + addon.upgrade_info.available`. In other words, a string concatenated with a boolean. This is an error. It might be worth configuring ESLint to flag errors in these situations:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": "error"
  }
}
```

The entire definition of `textValue` is confusing and complex, and involves repeated expressions. Perhaps ask the AI ​​to refactor it?

wesleybl commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonItem.tsx:
I think it's worth writing a test for this file. At least for its main component.

wesleybl commented on packages/volto/src/components/manage/Controlpanels/Addons/AddonsControlpanel.tsx:
```diff
@@ -0,0 +1,368 @@
+    defaultMessage: 'Error',
+  },
+  addonUpgraded: {
+    id: 'Addon upgraded succesfuly',
```
`succesfuly` -> `successfuly`

An `s` is missing.

In several places in this file.
